sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/catalog/InMemoryCatalog.scala

```diff
@@ -195,18 +195,31 @@ class InMemoryCatalog(
         throw new TableAlreadyExistsException(db = db, table = table)
       }
     } else {
-      if (tableDefinition.tableType == CatalogTableType.MANAGED) {
-        val dir = new Path(catalog(db).db.locationUri, table)
-        try {
-          val fs = dir.getFileSystem(hadoopConfig)
-          fs.mkdirs(dir)
-        } catch {
-          case e: IOException =>
-            throw new SparkException(s"Unable to create table $table as failed " +
-              s"to create its directory $dir", e)
+      val tableWithLocation = if (tableDefinition.tableType == CatalogTableType.MANAGED) {
+        val defaultTableLocation = new Path(catalog(db).db.locationUri, table)
+        // Ideally we can not create a managed table with location, but due to some limitations in
+        // [[CreateDataSourceTableAsSelectCommand]], we have to create the table directory and
+        // write out data before we create this table. We should handle this case and allow the
+        // table location to be pre-created, as long as it's same with the default table location.
+        if (tableDefinition.storage.locationUri.isDefined) {
+          val givenTableLocation = new Path(tableDefinition.storage.locationUri.get).toUri.toString
+          require(defaultTableLocation.toUri.toString == givenTableLocation)
```

Comment:
Can we unify the if and else branch? Should we also check whether defaultTableLocation directory path is created if tableDefinition.storage.locationUri is set?